JBOD support in KRaft mode (#108)
strimzi/proposals: 067-kraft-jbod-support.md

This proposal introduces JBOD support to KRaft-based Apache Kafka clusters powered by Strimzi.

## Current situation

Strimzi understands two different things under JBOD storage:

1. The `type: jbod` storage is a part of the `Kafka` custom resource API and allows users to define one or more disks in the form of an YAML list.
   The Strimzi storage type is not directly part of the configuration of the Apache Kafka nodes.
   It is only used as a basis to generate the value of the `log.dirs` configuration option that will contain one or more directories depending on the number of volumes defined in the `Kafka` CR.
   As a result, the Kafka nodes are not aware of the Strimzi storage type used in the `Kafka` CR.
   They are aware only of the number of directories / volumes used.
2. The actual use of the JBOD storage in the Apache Kafka nodes where multiple disks are used in each Kafka node.
   If the `log.dirs` configuration option contains more than one directory, JBOD storage is used.
   If it contains only one directory, JBOD storage is not used regardless of whether the `Kafka` CR uses `type: jbod` storage or not.

> Until version 3.7.0, Apache Kafka nodes do not support the use of JBOD storage.

Reviewer: in KRaft mode.

When running Apache Kafka in KRaft mode with Strimzi, we support using the `type: jbod` storage, but only with one data volume present.

## Motivation

In version 3.7.0, Apache Kafka introduces support for JBOD storage in KRaft as _early-access_.
It is expected to be GA in Apache Kafka 3.8.0.

> Use of JBOD storage one of the popular features among Strimzi (and Apache Kafka) users and we want to support it.

Reviewer: Use of JBOD storage **is** one of the popular features

But it will issue a warning when multiple volumes are specified for controller-only nodes.

### Dedicated metadata value

On brokers or on Kafka nodes with mixed roles, the volume / directory used for the metadata log will be used to also host regular partition-replicas.
In some situations, this might not be desired as the metadata log might compete with the regular Kafka logs for storage or performance.
Technically, it is possible to use a different data volume for the metadata log and for the regular Kafka logs.
However, it is not expected that this would be a common configuration because:
* Clusters sensitive to performance are expected to use dedicated controller-only nodes.
  And on the controllers-only nodes, there are no regular partition-replicas.
  So the metadata log has its own dedicated volume even without any special support for it.
* On broker-only nodes, the impact of shared volumes for data and metadata is expected to be small.
  And using a dedicated volume just for a metadata is not expected to be very efficient in terms of the utilization of the dedicated disk.

Based on the above, there will be no support for dedicated metadata log volumes.

> This should help us to safe unnecessary effort on the development and testing of this feature.

Reviewer: `safe` -> `save`

# JBOD support in KRaft mode

JBOD (Just a Bunch Of Disks) support in Apache Kafka means the ability to use multiple data directories in a single Kafka node to store the message logs.
Typically, the different data directories would represent different data volumes mounted on the Kafka node.
This proposal introduces JBOD support to KRaft-based Apache Kafka clusters powered by Strimzi.

## Current situation

> Strimzi understands two different things under JBOD storage:

Reviewer 2: ```suggestion
Strimzi distinguishes two aspects of JBOD storage:
```

1. The `type: jbod` storage is a part of the `Kafka` custom resource API and allows users to define one or more disks in the form of a YAML list.
   The Strimzi storage type is not directly part of the configuration of the Apache Kafka nodes.
   It is only used as a basis to generate the value of the `log.dirs` configuration option that will contain one or more directories depending on the number of volumes defined in the `Kafka` CR.
   As a result, the Kafka nodes are not aware of the Strimzi storage type used in the `Kafka` CR.
   They are aware only of the number of directories / volumes used.
2. The actual use of the JBOD storage in the Apache Kafka nodes where multiple disks are used in each Kafka node.
   If the `log.dirs` configuration option contains more than one directory, JBOD storage is used.
   If it contains only one directory, JBOD storage is not used regardless of whether the `Kafka` CR uses `type: jbod` storage or not.

> Until version 3.7.0, Apache Kafka nodes do not support the use of JBOD storage in KRaft mode.

Reviewer 2: ```suggestion
Prior to version 3.7.0, Apache Kafka nodes did not support JBOD storage in KRaft mode.
```

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  name: broker
  labels:
    strimzi.io/cluster: my-cluster
spec:
  replicas: 3
  roles:
    - broker
    storage:
      type: jbod
      volumes:
      - id: 0
        type: persistent-claim
        size: 100Gi
        deleteClaim: false
      - id: 1
        type: persistent-claim
        size: 100Gi
        deleteClaim: false
        kraftMetadata: shared
```

The reason why the `kraftMetadata` property is not just a boolean is to make it possible to extend the support in the future.
One such possibility would be to support dedicated metadata volumes, which is discussed in a separate section.

The `kraftMetadata` option will be ignored when used with ZooKeeper-based clusters.

### Kafka configuration

Currently, we only use the `log.dirs` option to configure the log directories.
With this proposal, the `log.dirs` option will be configured in the same way as present.
In addition to that, we will also configure the `metadata.log.dir` option that we do not use today.
This option will be configured only for KRaft clusters.
It will be set to the volume marked as `kraftMetadata: shared`.

> Or - if the flag is not set for any of the volumes - it will be set to the node with the lowest configured volume ID.

Reviewer 3: ```suggestion
Or - if the flag is not set for any of the volumes - it will be set to the volume with the lowest configured ID.
```

The currently used `metadata.log.dir` will be also stored in an annotation on the `StrimziPodSet` representing the node pool.
This annotation will be used to detect changes to the metadata log volume and used to log such events.
In the future, if necessary, it could also be used, for example, to prevent changes or to reassign the metadata log in a different way.
However, such functionality is not planned as part of this proposal.

### Changing the metadata log directory / volume

The volume / directory used for the metadata log might change in several situations:
* When a user changes the `kraftMetadata` flag and moves it to another volume
* When the volume originally used for the metadata is removed
* When a new volume with lower volume ID is added to the JBOD list and the `kraftMetadata` flag is not set

Changes to the `metadata.log.dir` always require rolling update of the affected Kafka nodes.
The change of the directory will be handled in the Kafka start-up scripts in our container image.
The scripts will do the following:
1) Try to find the existing metadata log
2) Compare the current location of the metadata log with the desired location

> 3) If the locations differ, the old metadata log will be deleted

Reviewer 4: Question: How would we delete the old metadata log while the same log dir contains other data logs? Only deleting the `__cluster_metadata` dir and `meta.properties` file?

Author: I don't remember exactly from when I was testing it - I think deleting the `__cluster_metadata-0` directory was enough.

Author: in any case, it would delete just the directories / files related to metadata and not the data related to regular message logs.

Reviewer 4: Yes, that's what I thought, and that should be fine as long as we don't touch other data log files.

Author: I updated the bullet point to make it more clear.

  And using a dedicated volume just for a metadata is not expected to be very efficient in terms of the utilization of the dedicated disk.

Based on the above, there will be no support for dedicated metadata log volumes.
This should help us to save unnecessary effort on the development and testing of this feature.

However, if needed, the support for it can be added later if we see the demand for it:
* Update the API and add a new valid value `dedicated` to the `kraftMetadata` field in the storage definition
* Configure this volume only as the `metadata.log.dir` and do not use it in `log.dirs`
* Add a validation to prevent changing of an existing volume used for regular logs to metadata-only log without all partition-replicas being moved to other volumes first.

### Example YAML files and documentation

The examples YAML files will be adapted to cover the API changes:
* A new KRaft JBOD example will be added with multiple disks and the `kraftMetadata` used for one of them
* The existing KRaft example with a single persistent volume will be adapted to add the `kraftMetadata: shared` option to the custom resource

> The documentation would suggest to Strimzi users to mark one of the volumes for being used for the metadata as that should help to make it more clear where the metadata are stored and might also help while helping users etc.

Reviewer 5: ```suggestion
The documentation would recommend that Strimzi users designate one of the volumes specifically for storing metadata. 
This would make the storage location for metadata clear, and might help with tasks such as maintenance.
```

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  name: broker
  labels:
    strimzi.io/cluster: my-cluster
spec:
  replicas: 3
  roles:
    - broker
    storage:
      type: jbod
      volumes:
      - id: 0
        type: persistent-claim
        size: 100Gi
        deleteClaim: false
      - id: 1
        type: persistent-claim
        size: 100Gi
        deleteClaim: false
        kraftMetadata: shared
```

The reason why the `kraftMetadata` property is not just a boolean is to make it possible to extend the support in the future.
One such possibility would be to support dedicated metadata volumes, which is discussed in a separate section.

The `kraftMetadata` option will be ignored when used with ZooKeeper-based clusters.

### Kafka configuration

Currently, we only use the `log.dirs` option to configure the log directories.
With this proposal, the `log.dirs` option will be configured in the same way as present.
In addition to that, we will also configure the `metadata.log.dir` option that we do not use today.
This option will be configured only for KRaft clusters.

> It will be set to the volume marked as `kraftMetadata: shared`.

Reviewer 6: It's not clear to me why the "shared" naming. It sounds like you are sharing the volume of the metadata with something else. But what? Isn't it dedicated to host the metadata only?

Reviewer 6: Actually I forgot that even brokers store metadata as "observers" so it makes sense.
